File: frontend/src/api/services.ts
import { z } from 'zod'
import { apiFetch } from './client'

const VoucherSpecSchema = z.discriminatedUnion('kind', [
  z.object({ kind: z.literal('Monthly') }),
  z.object({ kind: z.literal('Book'), amount: z.number(), duration: z.number() }),
])

export const ServiceSchema = z.object({
  id: z.number(),
  name: z.string(),
  description: z.string(),
  price: z.number(),
  voucher_spec: VoucherSpecSchema,
  external_service_id: z.number(),
})

export type Service = z.infer<typeof ServiceSchema>
export type VoucherSpec = z.infer<typeof VoucherSpecSchema>

const ServicesResponseSchema = z.object({
  data: z.array(ServiceSchema),
})

export async function listServices(): Promise<Service[]> {
  const raw = await apiFetch<unknown>('/api/services')
  return ServicesResponseSchema.parse(raw).data
}

File: src/routes/services.rs
use axum::{Json, extract::State};
use serde::Serialize;
use sqlx::FromRow;
use utoipa::ToSchema;

use crate::{AppState, auth::CurrentUser, domain::{Service, VoucherSpec}, error::AppError};

#[derive(FromRow)]
struct ServiceRow {
    id: i32,
    name: String,
    description: String,
    price: f64,
    kind: String,
    amount: Option<i32>,
    duration: Option<i32>,
    external_service_id: i32,
}

impl TryFrom<ServiceRow> for Service {
    type Error = AppError;

    fn try_from(row: ServiceRow) -> Result<Self, Self::Error> {
        let voucher_spec = match row.kind.as_str() {
            "Monthly" => VoucherSpec::Monthly,
            "Book" => VoucherSpec::Book {
                amount: row.amount.unwrap_or(1),
                duration: row.duration.unwrap_or(1),
            },
            _ => return Err(AppError::NotFound),
        };
        Ok(Service {
            id: row.id,
            name: row.name,
            description: row.description,
            price: row.price,
            voucher_spec,
            external_service_id: row.external_service_id,
        })
    }
}

#[derive(Serialize, ToSchema)]
pub struct ServicesResponse {
    pub data: Vec<Service>,
}

#[utoipa::path(
    get,
    path = "/services",
    tag = "Services",
    security(("bearer_auth" = [])),
    responses(
        (status = 200, description = "List of available services", body = ServicesResponse),
        (status = 401, description = "Unauthorized"),
    )
)]
pub async fn list_services(
    State(state): State<AppState>,
    _user: CurrentUser,
) -> Result<Json<ServicesResponse>, AppError> {
    let rows = sqlx::query_as::<_, ServiceRow>(
        r#"
        SELECT id, name, description, price, kind, amount, duration, external_service_id
        FROM portal_service
        WHERE is_available = true
        ORDER BY id
        "#,
    )
    .fetch_all(&state.db)
    .await?;

    let data = rows
        .into_iter()
        .map(Service::try_from)
        .collect::<Result<Vec<_>, _>>()?;

    Ok(Json(ServicesResponse { data }))
}
